File: core/src/ru/geekbrains/screen/GameScreen.java
package ru.geekbrains.screen;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.Texture;

import java.util.ArrayList;
import java.util.List;

import ru.geekbrains.spaceObjects.Asteroid;
import ru.geekbrains.base.Point;
import ru.geekbrains.spaceObjects.Bullet;
import ru.geekbrains.spaceObjects.SpaceShip;
import ru.geekbrains.base.Base2DScreen;

public class GameScreen extends Base2DScreen {

    Texture wallpaperImg;//, bulletImg, asteroidImg;

    Point dest;

    SpaceShip ship;
    List<Asteroid> asteroidList;

    int ASTEROID_QUANTITY = 7;

    @Override
    public void show() {
        super.show();

        ship = new SpaceShip(SCREEN_WIDTH / 2, 0);
        ship.setWidth(SCREEN_WIDTH / 100 * 10);
        ship.setHeight(SCREEN_HEIGHT / 100 * 15);

        initAsteroids();

        wallpaperImg = new Texture("space.jpg");
        dest = new Point(ship.getPosition());
    }
    private void initAsteroids() {
        asteroidList = new ArrayList<Asteroid>();
        for (int i = 0; i < ASTEROID_QUANTITY; i ++) {
            Asteroid asteroid = new Asteroid();
            asteroid.setWidth(SCREEN_WIDTH / 100 * 5);
            asteroid.setHeight(SCREEN_HEIGHT / 100 * 10);
            asteroid.setPosition(new Point(
                    Math.random() * 1024 % SCREEN_WIDTH,
                    SCREEN_HEIGHT + Math.random() * 1000 % SCREEN_HEIGHT, 0));
            asteroidList.add(i, asteroid);
        }
    }

    @Override
    public void render(float delta) {
        super.render(delta);
        Gdx.gl.glClearColor(0.5f, 0.2f, 0.3f, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
        drawGameObjects();

        ship.move(dest);
        moveSpaceShipBullets();
        ship.fire();
        moveAsteroids();

        checkDamage();
    }

    private void checkDamage() {
        // 1 TODO: 28.01.2019 Проверить урон, который был нанесен астероиду
        // 2 TODO: 28.01.2019 Удалить астероид при столкновении с кораблем
        // 3 TODO: 28.01.2019 Проверить урон, нанесенный кораблю
        // 4 TODO: 28.01.2019 Проверить урон, нанесенный кораблям врага
        for (Asteroid asteroid : asteroidList){
            if(ship.wasDamaged(asteroid)) {
                ship.setHealth(ship.getHealth() - asteroid.getDamage()); // ship damage (2)
                asteroid.resetAsteroid(SCREEN_WIDTH, SCREEN_HEIGHT);
            }
            for (Bullet bullet : ship.getBulletList()){
                if (asteroid.wasDamaged(bullet)) {
                    asteroid.setHealth(asteroid.getHealth() - bullet.getDamage());
                    bullet.resetBullet();
                    // check asteroid health
                    if (asteroid.getHealth() <= 0){
                        asteroid.resetAsteroid(SCREEN_WIDTH, SCREEN_HEIGHT);
                        ship.addExperience();
                    }
                }
            }

        }
        // check ship health
        if (ship.getHealth() <= 0) {
            ship.downgrade();
//            spaceShipImg = new Texture(ship.getOutfit());
        }

    }

    private void moveSpaceShipBullets() {
        for (Bullet b : ship.getBulletList()) {
            b.move(new Point(SCREEN_WIDTH, SCREEN_HEIGHT, 0));
        }
    }

    private void moveAsteroids() {
        for (Asteroid asteroid : asteroidList) {
            asteroid.move(new Point(SCREEN_WIDTH, SCREEN_HEIGHT, 0));
        }
    }

    private void drawGameObjects() {
        batch.begin();
        batch.draw(wallpaperImg,0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
        batch.end();
        ship.render(batch);
        for (Asteroid asteroid : asteroidList) {
            asteroid.render(batch);
        }

        for (Bullet bullet : ship.getBulletList()) {
            bullet.render(batch);
        }
    }

    @Override
    public void resize(int width, int height) {
        super.resize(width, height);
        ship.resize(SCREEN_WIDTH, SCREEN_HEIGHT);
        for (Bullet b : ship.getBulletList())
            b.resize(SCREEN_WIDTH, SCREEN_HEIGHT);
        for (Asteroid a : asteroidList)
            a.resize(SCREEN_WIDTH, SCREEN_HEIGHT);
    }

    @Override
    public void dispose() {
        wallpaperImg.dispose();
        batch.dispose();
        ship.dispose();
        for (Bullet b : ship.getBulletList())
            b.dispose();
        for (Asteroid a : asteroidList)
            a.dispose();

        super.dispose();
    }

    @Override
    public boolean touchDown(int screenX, int screenY, int pointer, int button) {
        dest = new Point(screenX, SCREEN_HEIGHT - screenY, 0);
        return super.touchDown(dest, pointer);
    }

}

File: core/src/ru/geekbrains/screen/GameScreen.kt
package ru.geekbrains.screen

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.audio.Music
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import ru.geekbrains.base.*
import ru.geekbrains.spaceObjects.SpaceShip
import java.io.*

class GameScreen : Base2DScreen() {

    private var wallpaperImg: Texture

    private var dest: Point
    private var ship: SpaceShip

    private var music: Music? = null
    
    private var memeManager: MemeManager
    
    var textureAtlas: TextureAtlas = TextureAtlas("meme_spaceship_sprite.txt")

    init {
        super.show()
        wallpaperImg = Texture("space.jpg")
        ship = SpaceShip()
        if (Gdx.files.local(".\\gameInstance.ser").exists()){
            restoreInstance()
            println("INSTANCE RESTORED")
        } else {
            ship = SpaceShip((SCREEN_WIDTH / 2).toDouble(), 0.0)
            ship.width = SCREEN_WIDTH / 100 * 10
            ship.height = SCREEN_HEIGHT / 100 * 15
        }
        dest = Point(ship.position)
        music = Gdx.audio.newMusic(Gdx.files.internal("shooting_stars.mp3"))
        music!!.isLooping = true
        music!!.play()

        memeManager = MemeManager()
        memeManager.memeFactory = MemeFactory(SCREEN_WIDTH.toDouble(), SCREEN_WIDTH.toDouble())
        memeManager.generateMemeList()
        
        memeManager.award = memeManager.memeFactory.getAward()
        
    }

    override fun render(delta: Float) {
        super.render(delta)
        Gdx.gl.glClearColor(0.5f, 0.2f, 0.3f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
        drawGameObjects()

        memeManager.move()
        ship.move()
        
        memeManager.checkDamage(ship)

        val updateWithoutDamageCounter = ship.checkDamage(memeManager.memeList)
        memeManager.checkAward(updateWithoutDamageCounter, batch)
    
        if (Gdx.input.isKeyJustPressed(Input.Keys.ESCAPE) || Gdx.input.isKeyJustPressed(Input.Keys.BACK))
            saveInstance()
    }
    
    private fun drawGameObjects() {
        batch.begin()
        batch.draw(wallpaperImg, 0f, 0f, SCREEN_WIDTH.toFloat(), SCREEN_HEIGHT.toFloat())
        batch.end()

        memeManager.render(batch)
        ship.render(batch)
        ship.fire()
    }

    private fun saveInstance(){
        var instance = GameInstance()
        instance.ship = this.ship
        instance.musicBeginningTimeSec = this.music!!.position.toDouble()
    
        try {
            var fileOut = FileOutputStream(".\\gameInstance.ser")
            var outStream = ObjectOutputStream(fileOut)
            outStream.writeObject(instance)
            outStream.close()
            fileOut.close()
            println("Serialized data is saved in \\gameInstance.ser")
        } catch (i: IOException) {
            i.printStackTrace()
        }
    }
    
    private fun restoreInstance() {
        var instance: GameInstance
        try {
            var fileIn = FileInputStream(".\\gameInstance.ser")
            var inStream = ObjectInputStream(fileIn)
            instance = inStream.readObject() as GameInstance
            inStream.close()
            fileIn.close()
        } catch (i: IOException) {
            i.printStackTrace()
            return
        } catch (c: ClassNotFoundException) {
            System.out.println("GameInstance class not found")
            c.printStackTrace()
            return
        }
    
        println(instance)
        ship = instance.ship
        ship.restore()
        println(ship)
    }
 
    override fun resize(width: Int, height: Int) {
        super.resize(width, height)
        ship.resize(SCREEN_WIDTH.toDouble(), SCREEN_HEIGHT.toDouble())
        memeManager.resize(SCREEN_WIDTH.toDouble(), SCREEN_HEIGHT.toDouble())
        memeManager.award.resize(SCREEN_WIDTH, SCREEN_HEIGHT)
        ship.bulletList.map { it -> it.resize(SCREEN_WIDTH.toDouble(), SCREEN_HEIGHT.toDouble()) }
    }

    override fun dispose() {
        wallpaperImg.dispose()
        batch.dispose()
        textureAtlas.dispose()

        super.dispose()
    }

    override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        dest = Point(screenX, SCREEN_HEIGHT - screenY, 0)
        ship.updateDirection(screenX.toDouble(), SCREEN_WIDTH.toDouble(), false)
        return super.touchDown(dest, pointer)
    }

    override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        ship.updateDirection(screenX.toDouble(), SCREEN_WIDTH.toDouble(), true)
        return super.touchUp(screenX, screenY, pointer, button)
    }

    override fun pause() {
        super.pause()
        music!!.pause()
    }

    override fun resume() {
        super.resume()
        music!!.play()
    }
}